design/decouple-ceph-version.md

```diff
@@ -0,0 +1,88 @@
+
+## Coupled (Legacy) Design
+The approach of embedding Ceph into the Rook image had several advantages that contributed to the design.
+- Simpler development and test matrix. A consistent version of Ceph is managed and there are no unstable Ceph bits running in the cluster.
+- Simpler upgrade path. There is only one version to worry about upgrading.
+
+The project is growing out of these requirements and we need to support some added complexity in order to get the benefits of the decoupled versions.
+
+## New Design
+
+There are two versions that will be specified independently in the cluster: the Rook version and the Ceph version.
+
+### Rook Version
+
+The Rook version is defined by the operator's container `image` tag. All Rook containers launched by the operator will also launch the same version of the Rook image.
+The full image name is an important part of the version. This allows the container to be loaded from a private repo if desired.
+
+In this example, the Rook version is `rook/ceph:v0.8.1`.
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: rook-ceph-operator
+spec:
+  template:
+    spec:
+      containers:
+      - name: rook-ceph-operator
+        image: rook/ceph:v0.8.1
+...
+```
+
+### Ceph Version
+
+The Ceph version is defined as the property `cephImage` in the Cluster CRD. All Ceph daemon containers launched by the Rook operator will use this image, including the mon, mgr,
+osd, rgw, and mds pods. The significance of this approach is that the Rook binary is not included in the daemon containers. All initialization performed by Rook to initialize and generate the Ceph config must be completed in an [init container](https://github.com/rook/rook/issues/2003). Once the Rook init containers complete their execution, the daemon container can run the Ceph image. The daemon container will no longer have Rook running.
+
+In the following Cluster CRD example, the Ceph version is Mimic 13.2.1.
+
+```yaml
+apiVersion: ceph.rook.io/v1beta1
+kind: Cluster
+metadata:
+  name: rook-ceph
+  namespace: rook-ceph
+spec:
+  cephImage: ceph/ceph:v13.2.1
+...
+```
+
+### Operator Requirements
+
+The operator needs to run the Ceph client tools to manage the cluster. For example, the `ceph` tool is needed for general Ceph configuration and status, `radosgw-admin` is required for managing an object store. Therefore, all the necessary client tools will still be included in the Rook image.
+
+The client tools are tested by the Ceph team to be backward and forward compatible by two versions. Therefore, the operator can support a version of Ceph up to two versions older than the client tools it contains. The operator will also be able to orchestrate upgrading to two versions newer of Ceph, in case there is a need to connect the operator to a newer version of a daemon before the operator is updated.
```

Comment:
I think it will be the other way around: the nautilus rook image will upgrade from luminous or mimic to nautilus, but not beyond, since it won't know anything about the specific upgrade procedures for O or P.

Reply:
Perhaps during development this makes it possible to experiment with upgrading to unreleased versions. But it isn't a customer scenario, so i'll reword to focus on upgrading versions that Rook understands.

Reply:
Yeah, we can have a "generic" upgrade path that just restarts daemons in order.  I think we should be *very* careful about letting users make use of that capability, though, since rook is venturing into unknown territory there.

Reply:
(I like the "AllowUnrecognizedVersion" proposal below.)

Reply:
I believe we will need the operator to support upgrading to one version of Ceph newer than the tools it contains. Here's the scenario:
- Rook releases 0.9 with the Mimic tools.
- Rook 0.9 can upgrade from luminous to mimic
- Rook 0.9 could also (mostly) support upgrade to nautilus since the latest nautilus builds could be tested for upgrade. Nautilus is likely to be released not too long after 0.9, or even around the same time. If any fixes are needed once Nautilus releases, Rook could have an 0.9 patch release to fully support it. An update of the mimic tools inside Rook would not be necessary.